### PR TITLE
[Test] Add ModalHelper Test

### PR DESCRIPTION
**Pull request by bblurock**

### CODE REVIEW

**Comment by bblurock** [comment]
[這個 Static](https://github.com/bblurock/windwalker-joomla-rad/blob/test-helper-ModalHelper/src/Helper/ModalHelper.php#L36) 讓我不得不跳掉兩個 method, 一個是 `modal()` 一個是 `renderModal()` [Link](https://github.com/bblurock/windwalker-joomla-rad/blob/test-helper-ModalHelper/src/Helper/ModalHelper.php#L76)

**Comment by bblurock** [comment]
然後看來是沒有任何可以 Mock/Stub static 的方法了

https://github.com/sebastianbergmann/phpunit-documentation/issues/77

**Comment by asika32764** [comment]
你可以用 Reflection 檢查 `JHtmlBootstrap::$loaded` 有沒有吃到東西就好

https://github.com/joomla/joomla-cms/blob/staging/libraries/cms/html/bootstrap.php#L265

**Comment by bblurock** [comment]
目前會出現的錯誤是這個

> ## Fatal error: Call to undefined method Windwalker\Console\Application\Console::getTemplate()
> 
> .........S.PHP Fatal error:  Call to undefined method Windwalker\Console\Application\Console::getTemplate() in /home/travis/build/joomla/libraries/cms/html/html.php on line 334
> PHP Stack trace:
> PHP   1. {main}() /home/travis/.phpenv/versions/5.3.29/bin/phpunit:0
> PHP   2. PHPUnit_TextUI_Command::main() /home/travis/.phpenv/versions/5.3.29/bin/phpunit:722
> PHP   3. PHPUnit_TextUI_Command->run() phar:///home/travis/.phpenv/versions/5.3.29/bin/phpunit/phpunit/TextUI/Command.php:104
> PHP   4. PHPUnit_TextUI_TestRunner->doRun() phar:///home/travis/.phpenv/versions/5.3.29/bin/phpunit/phpunit/TextUI/Command.php:152
> PHP   5. PHPUnit_Framework_TestSuite->run() phar:///home/travis/.phpenv/versions/5.3.29/bin/phpunit/phpunit/TextUI/TestRunner.php:398
> PHP   6. PHPUnit_Framework_TestSuite->run() phar:///home/travis/.phpenv/versions/5.3.29/bin/phpunit/phpunit/Framework/TestSuite.php:716
> PHP   7. PHPUnit_Framework_TestCase->run() phar:///home/travis/.phpenv/versions/5.3.29/bin/phpunit/phpunit/Framework/TestSuite.php:716
> PHP   8. PHPUnit_Framework_TestResult->run() phar:///home/travis/.phpenv/versions/5.3.29/bin/phpunit/phpunit/Framework/TestCase.php:693
> PHP   9. PHPUnit_Framework_TestCase->runBare() phar:///home/travis/.phpenv/versions/5.3.29/bin/phpunit/phpunit/Framework/TestResult.php:609
> PHP  10. PHPUnit_Framework_TestCase->runTest() phar:///home/travis/.phpenv/versions/5.3.29/bin/phpunit/phpunit/Framework/TestCase.php:737
> PHP  11. ReflectionMethod->invokeArgs() phar:///home/travis/.phpenv/versions/5.3.29/bin/phpunit/phpunit/Framework/TestCase.php:860
> PHP  12. Windwalker\Test\Helper\ModalHelperTest->testRenderModal() phar:///home/travis/.phpenv/versions/5.3.29/bin/phpunit/phpunit/Framework/TestCase.php:860
> PHP  13. Windwalker\Helper\ModalHelper::renderModal() /home/travis/build/joomla/libraries/windwalker/test/Helper/ModalHelperTest.php:137
> PHP  14. Windwalker\Helper\ModalHelper::modal() /home/travis/build/joomla/libraries/windwalker/src/Helper/ModalHelper.php:76
> PHP  15. JHtml::_() /home/travis/build/joomla/libraries/windwalker/src/Helper/ModalHelper.php:36
> PHP  16. JHtml::call() /home/travis/build/joomla/libraries/cms/html/html.php:138
> PHP  17. call_user_func_array:{/home/travis/build/joomla/libraries/cms/html/html.php:236}() /home/travis/build/joomla/libraries/cms/html/html.php:236
> PHP  18. JHtmlBootstrap::modal() /home/travis/build/joomla/libraries/cms/html/html.php:236
> PHP  19. JHtmlBootstrap::framework() /home/travis/build/joomla/libraries/cms/html/bootstrap.php:268
> PHP  20. JHtml::_() /home/travis/build/joomla/libraries/cms/html/bootstrap.php:230
> PHP  21. JHtml::call() /home/travis/build/joomla/libraries/cms/html/html.php:138
> PHP  22. call_user_func_array:{/home/travis/build/joomla/libraries/cms/html/html.php:236}() /home/travis/build/joomla/libraries/cms/html/html.php:236
> PHP  23. JHtmlJquery::framework() /home/travis/build/joomla/libraries/cms/html/html.php:236
> PHP  24. JHtml::_() /home/travis/build/joomla/libraries/cms/html/jquery.php:53
> PHP  25. JHtml::call() /home/travis/build/joomla/libraries/cms/html/html.php:138
> PHP  26. call_user_func_array:{/home/travis/build/joomla/libraries/cms/html/html.php:236}() /home/travis/build/joomla/libraries/cms/html/html.php:236
> PHP  27. JHtml::script() /home/travis/build/joomla/libraries/cms/html/html.php:236
> PHP  28. JHtml::includeRelativeFiles() /home/travis/build/joomla/libraries/cms/html/html.php:680

我後來發現在測試環境的時候  `JFactory::getApplication()` 拿到的是 console application `Windwalker\Console\Application\Console` 不是 cms application 所以不會有 `getTemplate()`

**Comment by asika32764** [comment]
參考 ComponentTest 裡面用 TestContainerHelper 切換 Application

**Comment by bblurock** [comment]
好的～大感謝!

**Comment by asika32764** [comment]
沒事要看一下其他人寫的 test

**Comment by asika32764** [comment]
See: https://github.com/smstw/windwalker-joomla-rad/pull/85

建議之後 rebase 到這個 PR 上

**Comment by bblurock** [comment]
Okay

File: src/Helper/ModalHelper.php [hunk 1]
```diff
@@ -48,14 +48,16 @@ public static function modal($selector, $option = array())
...
 		$onclick = isset($option['onclick']) ? " onclick=\"{$option['onclick']}\"" : '';
 		$icon    = JArrayHelper::getValue($option, 'icon', '');
 
-		$button = "<{$tag} data-toggle=\"modal\" data-target=\"#$selector\"{$id}{$class}{$onclick}>
-               <i class=\"{$icon}\" title=\"$title\"></i>
-                $title</{$tag}>";
+		$button = <<<HTML
+<{$tag} data-toggle="modal" data-target="#$selector"{$id}{$class}{$onclick}>
+	<i class="{$icon}" title="$title"></i>
+$title</{$tag}>
```
Comment (LeoOnTheEarth) [inline src/Helper/ModalHelper.php]: 這行的排版可以再調整一下
將 `$title` 與 `<i>` 對齊
`</{$tag}>` 移到下一行

File: test/Helper/ModalHelperTest.php [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	 * This method is called before a test is executed.
+	 *
+	 * @return void
+	 */
+	protected function setUp()
+	{
+		$app        = new TestApplication;
+		$app->input = new \JInput;
+
+		TestContainerHelper::setApplication($app);
```
Comment (LeoOnTheEarth) [inline test/Helper/ModalHelperTest.php]: 目前最新的 `dev-2.1` 不需要再做這樣的設定了
在 UnitTest 裏頭，預設是使用 `TestApplication`

File: test/Helper/ModalHelperTest.php [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+	public function testModal()
+	{
+		// Execute method
+		ModalHelper::modal('tag');
+
+		$bootstrap = new \ReflectionClass('JHtmlBootstrap');
+
+		// Load protected property
+		$loadedProp = $bootstrap->getProperty('loaded');
+		$loadedProp->setAccessible(true);
```
Comment (LeoOnTheEarth) [inline test/Helper/ModalHelperTest.php]: 上面需要換行

File: test/Helper/ModalHelperTest.php [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+		// Get value from protected property '$loaded'
+		$loaded = $loadedProp->getValue('JHtmlBootstrap');
+
+		// Test if JHtmlBootstrap::modal did executed
+		$this->assertEquals(true, isset($loaded['JHtmlBootstrap::modal']));
```
Comment (LeoOnTheEarth) [inline test/Helper/ModalHelperTest.php]: 可以用 `assertArrayHasKey`

File: test/Helper/ModalHelperTest.php [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+	 *
+	 * @covers Windwalker\Helper\ModalHelper::renderModal
+	 */
+	public function testRenderModal()
+	{
+		$bootstrap = new \ReflectionClass('JHtmlBootstrap');
+
+		// Load protected property
+		$loadedProp = $bootstrap->getProperty('loaded');
+		$loadedProp->setAccessible(true);
```
Comment (LeoOnTheEarth) [inline test/Helper/ModalHelperTest.php]: 上面需要換行

File: test/Helper/ModalHelperTest.php [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+<div class="modal-footer">
+    {$option['footer']}
+</div>
+</div>
+HTML;
+
+		$this->assertEquals($expected, ModalHelper::renderModal($selector, $option['content'], $option));
+
+		// Test if JHtmlBootstrap::modal did executed
+		$this->assertEquals(true, isset($loaded['JHtmlBootstrap::modal']));
```
Comment (LeoOnTheEarth) [inline test/Helper/ModalHelperTest.php]: 可以用 `assertArrayHasKey`

Comment (LeoOnTheEarth) [inline test/Helper/ModalHelperTest.php]: 另外因為 `JHtmlBootstrap::$loaded` 是共用的
在前一個 UnitTest 可能已經被 set 進去過
這邊的 assert 可能會有問題

參考一下我這邊的做法

https://github.com/LeoOnTheEarth/windwalker-joomla-rad/blob/757a06a6c28ff62584fc17f810910fe02fac202d/test/Script/ScriptManagerTest.php#L72-L117

在每個 test 裡都還原 `JHtmlBootstrap::$loaded` 原本的樣子會比較好

**Comment by LeoOnTheEarth** [comment]
`test/Helper/stubQuickaddForm.xml`
這個檔案看要不要移到 `test/Helper/Stub/stubQuickaddForm.xml`

**Comment by asika32764** [comment]
未完成的盡快處理，PR 積壓有點多

**Comment by bblurock** [comment]
Done.

File: test/Helper/ModalHelperTest.php [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+		$expectedForm = "
+<div class=\"control-group\" id=\"{$id}_title-wrap\">
+	<div class=\"control-label\">
+		<label id=\"{$id}_title-lbl\" for=\"{$id}_title\" class=\"required\">
+	stub_label<span class=\"star\">&#160;*</span></label>
+	</div>
+	<div class=\"controls\">
+		<input type=\"text\" name=\"{$id}[title]\" id=\"{$id}_title\" value=\"\" class=\"input-xlarge required\" required aria-required=\"true\" />
+	</div>
+</div>";
```
Comment (asika32764) [inline test/Helper/ModalHelperTest.php]: 這整段你用 Heredoc 會比較好處理

``` php
$html = <<<EOC
<html class'"{$class}">
EOC;
```

**Comment by asika32764** [comment]
加速處理

**Comment by bblurock** [comment]
改好了。

**Comment by asika32764** [comment]
找人 review merge 吧

**Comment by bblurock** [comment]
@gundamew 請幫我 review

**Comment by gundamew** [comment]
測試不過，我下的指令是：

``` bash
$ phpunit -v --debug --color=auto -c phpunit.xml --bootstrap test/bootstrap.php test/Helper/ModalHelperTest.php
```

結果是：

``` bash
Starting test 'Windwalker\Test\Helper\ModalHelperTest::testModal'.
E
Starting test 'Windwalker\Test\Helper\ModalHelperTest::testModalLink'.
.
Starting test 'Windwalker\Test\Helper\ModalHelperTest::testRenderModal'.
E
Starting test 'Windwalker\Test\Helper\ModalHelperTest::testGetQuickaddForm'.
E

There were 3 errors:

1) Windwalker\Test\Helper\ModalHelperTest::testModal
RuntimeException: Could not connect to MySQL.

./libraries/joomla/database/driver/mysql.php:83
./libraries/joomla/database/driver/mysql.php:140
./libraries/joomla/database/driver.php:1448
./libraries/cms/library/helper.php:148
./libraries/cms/library/helper.php:47
./libraries/cms/library/helper.php:90
./libraries/cms/version/version.php:177
./libraries/joomla/factory.php:736
./libraries/joomla/factory.php:216
./libraries/cms/html/html.php:701
./libraries/cms/html/html.php:236
./libraries/cms/html/html.php:138
./libraries/cms/html/jquery.php:53
./libraries/cms/html/html.php:236
./libraries/cms/html/html.php:138
./libraries/cms/html/bootstrap.php:230
./libraries/cms/html/bootstrap.php:268
./libraries/cms/html/html.php:236
./libraries/cms/html/html.php:138
./libraries/windwalker/src/Helper/ModalHelper.php:36
./libraries/windwalker/test/Helper/ModalHelperTest.php:87

2) Windwalker\Test\Helper\ModalHelperTest::testRenderModal
RuntimeException: Could not connect to MySQL.

./libraries/joomla/database/driver/mysql.php:83
./libraries/joomla/database/driver/mysql.php:140
./libraries/joomla/database/driver.php:1448
./libraries/cms/library/helper.php:148
./libraries/cms/library/helper.php:47
./libraries/cms/library/helper.php:90
./libraries/cms/version/version.php:177
./libraries/joomla/factory.php:736
./libraries/joomla/factory.php:216
./libraries/cms/html/html.php:701
./libraries/cms/html/html.php:236
./libraries/cms/html/html.php:104
./libraries/cms/html/jquery.php:53
./libraries/cms/html/html.php:236
./libraries/cms/html/html.php:104
./libraries/cms/html/bootstrap.php:230
./libraries/cms/html/bootstrap.php:268
./libraries/cms/html/html.php:236
./libraries/cms/html/html.php:104
./libraries/windwalker/src/Helper/ModalHelper.php:36
./libraries/windwalker/src/Helper/ModalHelper.php:77
./libraries/windwalker/test/Helper/ModalHelperTest.php:175

3) Windwalker\Test\Helper\ModalHelperTest::testGetQuickaddForm
RuntimeException: Could not connect to MySQL.

./libraries/joomla/database/driver/mysql.php:83
./libraries/joomla/database/driver/mysql.php:140
./libraries/joomla/database/driver.php:1448
./libraries/cms/library/helper.php:148
./libraries/cms/library/helper.php:47
./libraries/cms/library/helper.php:90
./libraries/cms/version/version.php:177
./libraries/joomla/factory.php:736
./libraries/joomla/factory.php:216
./libraries/cms/html/html.php:701
./libraries/cms/html/html.php:236
./libraries/cms/html/html.php:104
./libraries/cms/html/jquery.php:53
./libraries/cms/html/html.php:236
./libraries/cms/html/html.php:104
./libraries/joomla/form/fields/text.php:188
./libraries/joomla/form/field.php:393
./libraries/joomla/form/fields/text.php:73
./libraries/windwalker/src/Helper/ModalHelper.php:162
./libraries/windwalker/test/Helper/ModalHelperTest.php:206
```

是我這邊還要改 phpunit.xml 或其他設定嗎？

**Comment by asika32764** [comment]
你的 php CLI 無法連線 MySQL 造成的，請 @michael520 幫你檢查 Mysql.sock 的連結正不正確

**Comment by michael520** [comment]
mysql host 更改到 127.0.0.1 試試看

**Comment by asika32764** [comment]
@michael520 順便帶他改用 PHPStorm 來作測試，docs 網站上有教學

**Comment by gundamew** [comment]
改在 PhpStorm 底下跑就通過了，但是為什麼？

**Comment by asika32764** [comment]
簡單來說你用來跑 phpunit 的 php CLI 環境連不到 MySQL

如果你是用 XAMPP 之類的，有可能是 mysql.sock 連結的問題，要看了才知道

你執行看看

```
sudo mkdir /var/mysql
sudo ln -s /Applications/XAMPP/xamppfiles/var/mysql/mysql.sock /var/mysql/mysql.sock
```

**Comment by gundamew** [comment]
這個我有做過了

``` bash
$ ll /var/mysql
total 8
lrwxr-xr-x  1 root  wheel    51B  3 16 17:27 mysql.sock@ -> /Applications/XAMPP/xamppfiles/var/mysql/mysql.sock
```

---

另外現在跑我自己的的測試又不行了

> `Warning: require_once(PHPUnit/Runner/Version.php): failed to open stream: No such file or directory in ...`

依照「出現 PHPUnit/Runner/Version.php 無法載入的錯誤時」這一段描述的內容去改過 php.jar 裡的 code 然後：
1. 保留解壓縮後的 php 路徑，或
2. 重新包成 jar 檔案

結果是 PhpStorm 連 PHP 檔案都認不得了

目前正在找其他方法

**Comment by asika32764** [comment]
你現在用 phpStorm 8 還是 9?

**Comment by gundamew** [comment]
我現在是用 PhpStorm 8

**Comment by asika32764** [comment]
目前改用 9 就沒有問題了，8 的話就是要完全照著 .jar 那段修改，出現問題要自行 debug 一下

**Comment by gundamew** [comment]
NAS 存的 license key 不能用在 PhpStorm 9 上

PhpStorm 8 的問題搞不定

因為跑出來的結果一樣，所以我先用 ST 3 處理囉

**Comment by gundamew** [comment]
因為不是 mysql.sock 的問題，我乾脆重裝了開發環境，發現測試會失敗可能是因為 configuration.php 的 `$host` 沒設定成 127.0.0.1，以及當初安裝 Windwalker-RAD 時沒有乾淨地安裝

排除環境問題之後，除了 `testGetQuickaddForm` 之外，其他的測試都成功了

詳細的測試結果如下：

``` bash
PHPUnit 4.7.7 by Sebastian Bergmann and contributors.

Runtime:    PHP 5.5.23
Configuration:  /Users/bschen/Repositories/ihealth-schedule-unit-test/libraries/windwalker/phpunit.xml
Warning:    Deprecated configuration setting "strict" used


Starting test 'Windwalker\Test\Helper\ModalHelperTest::testModal'.
.
Starting test 'Windwalker\Test\Helper\ModalHelperTest::testModalLink'.
.
Starting test 'Windwalker\Test\Helper\ModalHelperTest::testRenderModal'.
.
Starting test 'Windwalker\Test\Helper\ModalHelperTest::testGetQuickaddForm'.
F

Time: 887 ms, Memory: 17.25Mb

There was 1 failure:

1) Windwalker\Test\Helper\ModalHelperTest::testGetQuickaddForm
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
 '<div class="control-group" id="stubQuickaddForm_title-wrap">
    <div class="control-label">
-       <label id="stubQuickaddForm_title-lbl" for="stubQuickaddForm_title" class="required">
-   stub_label<span class="star">&#160;*</span></label>
+       <label id="stubQuickaddForm_title-lbl" for="stubQuickaddForm_title" class=" required">stub_label<span class="star">&#160;*</span></label>
    </div>
    <div class="controls">
-       <input type="text" name="stubQuickaddForm[title]" id="stubQuickaddForm_title" value="" class="input-xlarge required" required aria-required="true" />
+       <input type="text" name="stubQuickaddForm[title]" id="stubQuickaddForm_title" value="" class="input-xlarge" required aria-required="true" />
    </div>
 </div>'

/Users/bschen/Repositories/ihealth-schedule-unit-test/libraries/windwalker/test/Helper/ModalHelperTest.php:206

FAILURES!
Tests: 4, Assertions: 6, Failures: 1.
```

在 CLI 的測試結果和在 PhpStorm 的測試結果一樣

**Comment by asika32764** [comment]
我印象中只吃 localhost 有一種可能是因為跑得是 Mac 內建的 mysql 或者 XAMPP 的 mysql 沒有正常的啟動

我之前出現這種狀況是用 CLI kill 掉 mysql 的 process 再重新用 XAMPP 啟動就好了

你可以檢查一下在 CLI 下的 mysql 是否與 Joomla 用的 mysql 儲存內容一樣，如果一樣，應該要可以跟 Joomla 一樣吃 localhost，不能的話就表示 run 到另一個 mysql 去了

**Comment by gundamew** [comment]
重新安裝 XAMPP 的時候也有遇到這類問題：

- 重啟 XAMPP apache/mysql 服務失敗，訊息顯示已有服務在跑
- `ps aux` 和 Activity Monitor 沒看到東西
- 重開機之後，正常啟動 XAMPP apache/mysql 服務

`$ /Applications/XAMPP/xamppfiles/bin/mysql` 看到的東西也跟 phpMyAdmin 看到的東西不同

**Pull request merged**